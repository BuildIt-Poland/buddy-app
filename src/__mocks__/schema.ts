export enum UserRole {
  Newbie = 'NEWBIE',
  Buddy = 'BUDDY',
}

export enum TaskStatus {
  Completed = 'COMPLETED',
  Uncompleted = 'UNCOMPLETED',
}

export default {
  UserRole,
  TaskStatus,
};
