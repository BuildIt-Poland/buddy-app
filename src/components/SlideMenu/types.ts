import {
  Buddy,
  MutationDeleteBuddyArgs,
  MutationDeleteNewbieArgs,
  Newbie,
} from 'buddy-app-schema';

export type SlideMenuProps = {
  isMenuVisible: boolean;
  onClose: () => void;
};

export type BasicDetailsParams = MutationDeleteNewbieArgs | MutationDeleteBuddyArgs;

export type UserBasicDetails = {
  [key: string]: Partial<Newbie> | Partial<Buddy>;
};
