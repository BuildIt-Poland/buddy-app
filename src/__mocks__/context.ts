import { AuthContextData } from 'context/AuthStore';

export const authContext = (context?: Partial<AuthContextData>) => ({
  data: { role: 'BUDDY', token: 'dummy-token', userId: '1' },
  loading: false,
  error: null,
  isAuthenticated: true,
  login: jest.fn(),
  logout: jest.fn(),
  ...context,
});
