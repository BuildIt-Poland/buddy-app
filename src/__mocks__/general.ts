import { ROUTES } from 'shared/routes';

export const mockLocation = (path: string = ROUTES.BASE) => ({
  key: 'utwyk7',
  pathname: path,
});

export const mockSchemaTypes = (types: any = {}) => ({
  UserRole: {
    Newbie: 'NEWBIE',
    Buddy: 'BUDDY',
  },
  TaskStatus: {
    Completed: 'COMPLETED',
    Uncompleted: 'UNCOMPLETED',
  },
  ...types,
});
