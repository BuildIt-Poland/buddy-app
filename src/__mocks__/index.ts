export * from './general';
export * from './graphql';
export * from './schema';
