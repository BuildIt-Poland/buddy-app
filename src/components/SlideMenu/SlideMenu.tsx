import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Drawer from '@material-ui/core/Drawer';
import Divider from '@material-ui/core/Divider';
import CloseIcon from '@material-ui/icons/Close';
import Box from '@material-ui/core/Box';
import IconButton from '@material-ui/core/IconButton';
import { useQuery } from '@apollo/react-hooks';
import { useHistory } from 'react-router-dom';
import {
  BUDDY_BASIC_DETAILS,
  NEWBIE_BASIC_DETAILS,
} from 'graphql/contact-details.graphql';
import NewbiesMenuSection from 'components/NewbiesMenuSection';
import UserMenuDetails from 'components/UserMenuDetails';
import UserMenuSettings from 'components/UserMenuSettings';
import BuddyMenuSection from 'components/BuddyMenuSection';
import { isBuddy, isNewbie } from 'utils';
import theme from 'styles/theme';
import { ROUTES } from 'shared/routes';
import { Buddy, Newbie, UserRole } from 'buddy-app-schema';
import { BasicDetailsParams, SlideMenuProps, UserBasicDetails } from './types';

const useStyles = makeStyles({
  list: {
    width: '35rem',
    display: 'flex',
    flexDirection: 'column',
  },
  closeBtn: {
    alignSelf: 'flex-end',
    zIndex: theme.zIndex.base,
  },
});

const SlideMenu: React.FC<SlideMenuProps> = props => {
  const history = useHistory();
  const { list, closeBtn } = useStyles();
  const { isMenuVisible, onClose } = props;

  // FIXME: Get role from graphql
  const id = '';
  const role = UserRole.Newbie;

  const getQueryByRole = (role: UserRole, id: string) => {
    if (isBuddy(role)) {
      return { query: BUDDY_BASIC_DETAILS, variables: { buddyId: id } };
    } else if (isNewbie(role)) {
      return { query: NEWBIE_BASIC_DETAILS, variables: { newbieId: id } };
    }
  };

  const selectNewbie = (id: string) => {
    history.push(ROUTES.BUDDY_TASKS_LIST.replace(':newbieId', id));
    onClose();
  };

  const selectBuddy = (id: string) => {
    history.push(ROUTES.NEWBIE_BUDDY_DETAILS.replace(':buddyId', id));
    onClose();
  };

  const { query, variables } = getQueryByRole(role, id) || {};
  const { data } = useQuery<UserBasicDetails, BasicDetailsParams>(query, {
    variables,
  });
  const user = data && data[role.toLowerCase()];

  return user ? (
    <Drawer open={isMenuVisible} onClose={onClose}>
      <Box className={list} role='presentation'>
        <IconButton className={closeBtn} onClick={onClose}>
          <CloseIcon />
        </IconButton>
        <UserMenuDetails user={user} />
        <Divider />
        {isBuddy(role) && (
          <NewbiesMenuSection
            newbies={(user as Partial<Buddy>).newbies || []}
            onSelect={selectNewbie}
          />
        )}
        {isNewbie(role) && (
          <BuddyMenuSection
            buddy={(user as Partial<Newbie>).buddy || {}}
            onSelect={selectBuddy}
          />
        )}
        <Divider />
        <UserMenuSettings
          allowPushedNotifications={user.allowPushedNotifications || true}
          updatePushNotificationsSettings={() => {}}
        />
      </Box>
    </Drawer>
  ) : null;
};

export default SlideMenu;
