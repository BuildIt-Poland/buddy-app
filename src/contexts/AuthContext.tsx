import React from 'react';
import { AuthPayload } from 'buddy-app-schema';

export interface State {
  loading: boolean;
  error: any;
  data: AuthPayload;
  isAuthenticated: boolean;
  isAppBootstrapped: boolean;
}

export interface AuthContextData extends State {
  login: (email: string, password: string) => void;
  logout: () => void;
}

export const defaultState: State = {
  loading: false,
  error: null,
  data: {} as AuthPayload,
  isAuthenticated: false,
  isAppBootstrapped: false,
};

const defaultContext: AuthContextData = {
  ...defaultState,
  login: () => null,
  logout: () => null,
};

const AuthContext = React.createContext<AuthContextData>(defaultContext);

export default AuthContext;
