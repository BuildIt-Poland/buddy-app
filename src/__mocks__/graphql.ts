import { LOGIN_MUTATION } from 'graphql/login.graphql';
import { BUDDY_MENU_DETAILS, NEWBIE_MENU_DETAILS } from 'graphql/user-menu.graphql';
import { AVATAR_HEADER } from 'graphql/avatar-header.graphql';
import { NEWBIE_SELECT } from 'graphql/newbie-select.graphql';
import { TASK_DETAILS } from 'graphql/task-details.graphql';
import { GraphQLError } from 'graphql';
import {
  MutationLoginArgs,
  AuthPayload,
  TaskStatus,
  Newbie,
  QueryNewbieArgs,
  QueryBuddyArgs,
  UserRole,
} from 'buddy-app-schema';
import {
  NEWBIE_CONTACT_DETAILS,
  BUDDY_CONTACT_DETAILS,
} from 'graphql/contact-details.graphql';
import { getBasicUserDetailsMock } from './general';
import { authContext } from './context';

export const loginSuccessMock = (
  variables?: Partial<MutationLoginArgs>,
  data?: Partial<AuthPayload>
) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  result: {
    data: {
      login: {
        ...authContext().data,
        ...data,
      },
    },
  },
});

export const loginFailedMock = (variables?: Partial<MutationLoginArgs>) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  result: {
    errors: [new GraphQLError('No such user found')],
  },
});

export const loginNoNetworkMock = (variables?: Partial<MutationLoginArgs>) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  error: new Error('Network error'),
});

export const NewbieAvatarDetails = (variables?: QueryNewbieArgs) => ({
  request: {
    query: AVATAR_HEADER,
    variables,
  },
  result: {
    data: {
      newbie: {
        name: 'Tom Hanks',
        position: 'front-end',
        photo: 'some-url',
        tasksInfo: {
          buddyProgress: 0.54,
        },
      },
    },
  },
});

export const tasksResponse = (filter?: TaskStatus) => {
  const tasks = [
    {
      id: '1',
      title: 'Call you head manager',
      description: '',
      createdAt: '',
      implementationDate: '',
      status: TaskStatus.Uncompleted,
      newbie: {} as Newbie,
    },
    {
      id: '2',
      title: 'Write you main manager',
      description: '',
      createdAt: '',
      implementationDate: '',
      status: TaskStatus.Completed,
      newbie: {} as Newbie,
    },
  ];

  switch (filter) {
    case TaskStatus.Completed:
      return [tasks[0]];
    case TaskStatus.Uncompleted:
      return [tasks[1]];
    default:
      return tasks;
  }
};

export const buddyMenuDetails = (variables?: Partial<QueryBuddyArgs>) => [
  {
    request: {
      query: BUDDY_MENU_DETAILS,
      variables,
    },
    result: {
      data: {
        buddy: {
          name: 'Tom Hanks',
          email: 'tom@wipro.com',
          allowPushedNotifications: true,
          photo: 'buddy-photo-url',
          newbies: [
            {
              id: 'id',
              name: 'Sandra Bullock',
              photo: 'newbie-photo-url',
            },
          ],
        },
      },
    },
  },
];

export const newbieMenuDetails = (variables?: Partial<QueryNewbieArgs>) => [
  {
    request: {
      query: NEWBIE_MENU_DETAILS,
      variables,
    },
    result: {
      data: {
        newbie: {
          name: 'Sandra Bullock',
          email: 'doejoe@wipro.com',
          allowPushedNotifications: true,
          photo: 'newbie-photo-url',
          buddy: {
            id: 'id',
            name: 'Tom Hanks',
            photo: 'newbie-photo-url',
          },
        },
      },
    },
  },
];

export const newbieContactDetails = (variables?: Partial<QueryNewbieArgs>) => [
  {
    request: {
      query: NEWBIE_CONTACT_DETAILS,
      variables,
    },
    result: {
      data: {
        newbie: getBasicUserDetailsMock(UserRole.Newbie),
      },
    },
  },
];

export const buddyContactDetails = (variables?: Partial<QueryBuddyArgs>) => [
  {
    request: {
      query: BUDDY_CONTACT_DETAILS,
      variables,
    },
    result: {
      data: {
        buddy: getBasicUserDetailsMock(UserRole.Buddy),
      },
    },
  },
];

export const newbieSelectMock = [
  {
    request: {
      query: NEWBIE_SELECT,
      variables: {
        buddyId: 'ck17sl83c9gya0b17dcvttzm4',
      },
    },
    result: {
      data: {
        buddy: {
          id: 'ck17sl83c9gya0b17dcvttzm4',
          name: 'Dummy',
          role: 'BUDDY',
          photo: null,
          newbies: [
            {
              id: 'ck17svulh9k2k0b17j31ansfk',
              photo: null,
              name: 'Test 2',
              startDate: null,
              tasksInfo: { buddyProgress: 0.5 },
            },
            {
              id: 'ck17swp2m9kcv0b17we0ibrdn',
              photo: null,
              name: 'Test 1',
              startDate: null,
              tasksInfo: { buddyProgress: 0.6 },
            },
          ],
        },
      },
    },
  },
];

export const taskDetailsMock = [
  {
    request: {
      query: TASK_DETAILS,
      variables: {
        taskId: '1',
      },
    },
    result: {
      data: {
        task: {
          title: 'New task',
          description: 'New task',
          status: TaskStatus.Uncompleted,
        },
      },
    },
  },
];
