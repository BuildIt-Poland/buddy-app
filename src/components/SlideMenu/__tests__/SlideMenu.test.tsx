import React from 'react';
import { create, act } from 'react-test-renderer';
import SlideMenu from 'components/SlideMenu';
import { MockedProvider } from '@apollo/react-testing';
import { MemoryRouter, Route } from 'react-router';
import { authContext, buddyUserMenuDetails, mockSchemaTypes } from '__mocks__';
import AuthContext from 'contexts/AuthContext';
import waitForExpect from 'wait-for-expect';

jest.mock('buddy-app-schema', () => ({
  UserRole: {
    Newbie: 'NEWBIE',
    Buddy: 'BUDDY',
  },
}));
jest.mock('components/UserMenuDetails/UserMenuDetails', () => 'UserMenuDetails');
jest.mock('components/BuddyMenuSection/BuddyMenuSection', () => 'BuddyMenuSection');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');
jest.mock('@material-ui/icons/Close', () => 'CloseIcon');
jest.mock('@material-ui/core/Box', () => 'Box');
jest.mock('@material-ui/core/Drawer', () => 'Drawer');
jest.mock('@material-ui/core/Divider', () => 'Divider');

jest.mock(
  'components/NewbiesMenuSection/NewbiesMenuSection',
  () => 'NewbiesMenuSection'
);
jest.mock('components/UserMenuSettings/UserMenuSettings', () => 'UserMenuSettings');
jest.mock('components/UserMenuSettings/UserMenuSettings', () => 'UserMenuSettings');

const mockedContext = {
  data: {
    userId: 'id',
    token: 'token',
    role: mockSchemaTypes().UserRole.Buddy,
  },
};

const context: any = authContext(mockedContext);

describe('SlideMenu component', () => {
  const path = '/buddy/newbies';
  it('renders correctly', async () => {
    const onCloseMock = jest.fn();
    const tree = create(
      <MockedProvider
        mocks={[buddyUserMenuDetails({ buddyId: '1234' })]}
        addTypename={false}>
        <MemoryRouter initialEntries={[path]}>
          <Route path={path}>
            <AuthContext.Provider value={context}>
              <SlideMenu isMenuVisible={false} onClose={onCloseMock} />
            </AuthContext.Provider>
          </Route>
        </MemoryRouter>
      </MockedProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
    await act(async () => {
      await waitForExpect(() => {
        expect(tree).toMatchSnapshot();
      });
    });
  });
});
