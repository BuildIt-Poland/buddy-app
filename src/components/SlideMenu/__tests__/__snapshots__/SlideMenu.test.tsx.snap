// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlideMenu component renders correctly 1`] = `
<Drawer
  onClose={[MockFunction]}
  open={false}
/>
`;

exports[`SlideMenu component renders correctly 2`] = `
<Drawer
  onClose={[MockFunction]}
  open={false}
/>
`;
